--- pulp-sdk/configs/common.sh	2023-10-04 18:05:49.145551892 +0200
+++ pulp-sdk-mod/configs/common.sh	2023-10-04 18:00:38.780346436 +0200
@@ -29,6 +29,7 @@
 
 # Gapy
 export PATH=$PULP_SDK_HOME/tools/gapy:$PATH
+export PYTHONPATH=$PULP_SDK_HOME/tools/gap-configs/python:$PYTHONPATH
 export PYTHONPATH=$PULP_SDK_HOME/tools/gapy:$PYTHONPATH
 export PULP_CONFIGS_PATH=$PULP_SDK_HOME/tools/gap-configs/configs:$PULP_CONFIGS_PATH
 
