--- nemo/transf/export.py	2023-10-03 20:31:16.159056639 +0200
+++ nemo-mod/transf/export.py	2023-10-03 20:34:42.645857574 +0200
@@ -51,6 +51,12 @@
                 d[n]['W_scale'] = m.W_precision.get_scale()
         except AttributeError:
             pass
+        try:
+            d[n]['b_bits']  = m.b_precision.get_bits()
+            if not "PACT" in m.__class__.__name__:
+                d[n]['b_scale'] = m.b_precision.get_scale()
+        except AttributeError:
+            pass
         if len(d[n].keys()) == 0 or n == "":
             d.pop(n, None)
     return d
