--- nemo/graph.py	2023-10-03 20:26:14.201885400 +0200
+++ nemo-mod/graph.py	2023-10-03 20:19:22.534289311 +0200
@@ -132,7 +132,7 @@
                 try:
                     graph, _params_dict, _torch_out = torch.onnx.utils._model_to_graph(module, dummy_input, propagate=True, _retain_param_name=True)
                 except TypeError:
-                    graph, _params_dict, _torch_out = torch.onnx.utils._model_to_graph(module, dummy_input, _retain_param_name=True)
+                    graph, _params_dict, _torch_out = torch.onnx.utils._model_to_graph(module, dummy_input)
         input_dict = {}
         output_dict = {}
         self.non_unique_names_dict = {}
